### PR TITLE
[WIP] Enable tests that use DataLoader with multiple workers on Windows

### PR DESCRIPTION
**Pull request by yf225**
The CUDA OOM issue was fixed by #5585

### CODE REVIEW

**Comment by yf225** [comment]
@peterjc123 `test_batch_sampler` seems to fail in https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-test/3447/console

**Comment by peterjc123** [comment]
@yf225 I guess there's another issue in CUDA.

**Comment by peterjc123** [comment]
Well, I've tested using my own machine. Actually, there was a peak when a certain test was executed. I guess it's test_batch_sampler. Except this, the others are fine.
image

**Comment by peterjc123** [comment]
@yf225 Now I suspect it's `spawn` that caused this. Because the import part of `test_dataloader.py` is not protected by `if __name__ == '__main__':`, the `TEST_CUDNN` or `THCMalloc` in `common_nn` will be executed multiple times at the same time, which may be the cause of this issue.

File: test/test_dataloader.py [hunk 1]
```diff
@@ -13,7 +13,8 @@
 from torch.utils.data.dataset import random_split
 from torch.utils.data.dataloader import default_collate, ExceptionWrapper
 from common import TestCase, run_tests, TEST_NUMPY, IS_WINDOWS
-from common_nn import TEST_CUDA
+if __name__ == '__main__':
```
Comment (peterjc123) [inline test/test_dataloader.py]: Should add a default value for `TEST_CUDA` because it's used below.

Comment (yf225) [reply test/test_dataloader.py]: Would it be okay to put `TEST_CUDA = False` as default value here?

Comment (peterjc123) [reply test/test_dataloader.py]: I think it's okay.

**Comment by peterjc123** [comment]
In `test_utils.py`, I think ` HAS_CUDA = torch.cuda.is_available()` should be protected and the `HAS_CUDA` is not fully used in this script.

**Comment by peterjc123** [comment]
According to my test, `load_lua` should also be protected. But I don't know why.

**Comment by apaszke** [comment]
What is wrong with calling `is_available()` on Windows? It's not doing anything special

**Comment by peterjc123** [comment]
@apaszke Yes, you are right. But `load_lua` should be guarded. There seems to be some memory allocation in it.

**Comment by apaszke** [comment]
That is true, but why is it an issue? How many processes are spawned by the tests? Each of them should consume at most 300MB on the GPU, so I'd assume having 1GB available would be enough.

**Comment by peterjc123** [comment]
Although there're only 2 workers, I guess the child processes does not shutdown directly after the single test is over so the GPU memory is used up.

**Comment by apaszke** [comment]
Can't we force terminate the child processes?

**Comment by yf225** [comment]
@peterjc123 Seems like we haven't got the `test_cuda` in `TestBottleneck` fix right yet: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-test/3535/consoleFull

**Comment by peterjc123** [comment]
How weird it is! This test does not take even more than 1 GB in my machine. How can it run out of 8G of GPU on the test machine?

**Comment by yf225** [comment]
@peterjc123 Were you able to reproduce it on the CI dev machine?

**Comment by peterjc123** [comment]
@yf225 I cannot reproduce it on the dev machine either.

**Comment by peterjc123** [comment]
And another related issue #6288.

**Comment by yf225** [comment]
https://github.com/pytorch/pytorch/issues/5736 looks related in terms of killing child processes after parent dies. Working on a fix now.

**Comment by yf225** [comment]
Closing this in favor of https://github.com/pytorch/pytorch/pull/6745/.

**Pull request closed without merging**